src/Operation/ProcedureOperation.php

```diff
@@ -0,0 +1,187 @@
+     * @param string      $operation
+     * @param bool        $deterministic
+     * @param string      $readsSqlData
+     * @param array       $inParameterOperations
+     * @param array       $outParameterOperations
+     * @param string|null $body
+     */
+    public function __construct(
+        string $name,
+        string $operation,
+        bool $deterministic = false,
+        string $readsSqlData = 'READS SQL DATA',
+        array $inParameterOperations = [],
+        array $outParameterOperations = [],
+        string $body = null
+    ) {
+        $this->name = $name;
+        $this->operation = $operation;
+        $this->deterministic = $deterministic;
+        $this->readsSqlData = $readsSqlData;
+        $this->inParameterOperations = $inParameterOperations;
+        $this->outParameterOperations = $outParameterOperations;
+        $this->body = $body;
+    }
+
+    /**
+     * Returns the reverse of the operation.
+     *
+     * @param ReversibleOperationInterface|null $originalOperation
+     * @return ReversibleOperationInterface|null
+     */
+    public function reverse(?ReversibleOperationInterface $originalOperation = null): ?ReversibleOperationInterface
+    {
+        /* @var ProcedureOperation $originalOperation */
+        if ($originalOperation && $originalOperation->getName() !== $this->name) {
+            throw new InvalidArgumentException('Previous operations must apply to the same procedure.');
+        }
+
+        if ($this->getOperation() === ProcedureOperation::CREATE) {
+            return new ProcedureOperation(
+                $this->getName(),
+                ProcedureOperation::DROP
+            );
+        }
+
+        return $originalOperation;
+    }
+
+    /**
+     * Returns a new operation by applying another operation.
+     *
+     * @param ReducibleOperationInterface $operation
+     * @return ReducibleOperationInterface|null
+     */
+    public function apply(ReducibleOperationInterface $operation): ?ReducibleOperationInterface
+    {
+        /* @var ProcedureOperation $operation */
+        if ($operation->getName() !== $this->getName()) {
+            throw new InvalidArgumentException('Cannot apply operations for a different procedure.');
+        }
+
+        if ($this->getOperation() === self::DROP) {
+            throw new InvalidArgumentException('Cannot apply further operations to a dropped procedure.');
+        }
+
+        if ($this->getOperation() === self::CREATE) {
+            if ($operation->operation === self::CREATE) {
+                throw new InvalidArgumentException('Cannot recreate an existing procedure.');
+            }
+
+            // Skip creation of procedures that will be dropped
+            if ($operation->getOperation() === self::DROP) {
+                return null;
+            }
+        } else if ($operation->getOperation() === self::DROP) {
+            // Skip modification of procedures that will be dropped
+            return $operation;
```

Comment:
Judging by the direction we've taken with other types of operations, this should result in an exception being thrown.

Reply:
Any thoughts on this one? I don't think there's any scenario where altering a dropped procedure should be possible.

It could be recreated of course, but then we'd want to keep the DROP and subsequent CREATE as separate operations rather than trying to calculate an ALTER from the two and past history. The library is probably not in a place where this would be possible yet, and for our use cases it's unlikely to come up.

Happy to approve once we've made a decision here.